### DISCUSSION TITLE
cleanest / most `{targets}`-style way to export data resulting from pipeline

### DISCUSSION THREAD

**Discussion by januz** [ropensci/targets#965] [General]
We're using targets to run daily jobs that generates data for use in dashboards. For the dashboards, we just use the `repository = "aws"` option and read the data from the `_targets/` directory in the S3 bucket. But now I would like to additionally write some data resulting from the pipeline to another S3 bucket where it will be picked up by a different processing pipeline (that shouldn't have access to the S3 bucket that the main pipeline writes to).

I can do this in several ways but wonder what would be considered the best way / the style most in line with `{targets}`. Currently I was thinking of 

- either writing the data to S3 as a side effect of one of the functions that create targets in the pipeline
- or having an additional target that just executes the function writing data to S3

Both don't feel super "clean"

- the former feels as though it's not explicit enough
- the latter feels odd because the target's value would just be `NULL`

Is there a specific `targets/tarchetype` way to deal with situations like these where outputs of the pipeline are stored outside of `{targets}` purview?

**Comment by wlandau** [comment]
The most reproducible `targets`-compatible way I can think of at the moment is to set a different bucket and prefix for that target. Different targets can have different buckets.

```r
tar_target(
  name = external_output,
  command = get_external_output(upstream_target),
  resources = tar_resources(
    aws = tar_resources_aws(
      bucket = "different_bucket_just_for_external_output",
      prefix = "wherever/the/output/goes/",
      region = "region_of_other_bucket"
    )
  )
)
```

**Comment by januz** [comment]
@wlandau That looks like a great approach! Is there a way that I can specify a pair of S3 access credentials (access key and secret) within `tar_resources_aws()` that is different from the one that is used for the main pipeline? Thanks so much!

**Reply by wlandau** [reply]
The `...` argument of `tar_resources_aws()` allows you to pass named arguments to [`paws`](https://paws-r.github.io) functions. Hopefully that will allow you to pass a different set of credentials.

**Reply by januz** [reply]
Apparently I was using an older version of `{targets}` that didn't have the `...` argument yet. After updating, I see it in the documentation. Sorry for bothering you with the question!

**Reply by januz** [reply]
Hmm, I was expecting that I could pass `key` and `secret `arguments to `...` but I'm getting this error message:

```
The named arguments in ... must not be any of "bucket", "Bucket", "key", "Key", "prefix", "region", "part_size", "endpoint", "version", "VersionId", "body", "Body", "metadata", "Metadata", "UploadId", "MultipartUpload", or "PartNumber".
```

**Reply by wlandau** [reply]
Now that I am checking, I do see that arguments `access_key_id`, `secret_access_key`, and `session_token` are part of [`paws::s3()`](https://paws-r.github.io/docs/s3/) and that `targets` does not expose these arguments. On reflection, I think this is the right call. The access key ID and secret access key are so sensitive that `targets` should not keep them in memory. If these keys become `tar_resources_aws()` arguments, then they will become part of a target object and possibly sent over a network to a worker as part of [high-performance computing](https://books.ropensci.org/targets), and even TLS encryption is not always a perfect defense against man in the middle attacks. So for your case, you could run the pipeline in different stages with different AWS environment variables in the project's `.Renviron` file.

**Reply by januz** [reply]
@wlandau Are you saying that, even if I would would use `Sys.getenv()` in the call, the key/secret would become part of the `targets` object?

Would you mind pointing me to a resource on how to run the pipeline in different stages with different environment variables? 

Thanks so much for your help!

**Reply by wlandau** [reply]
> even if I would would use Sys.getenv() in the call, the key/secret would become part of the targets object?

No, `targets` does not know `access_key_id`, `secret_access_key`, or `session_token`. What I meant is that if I were to expose these as arguments to `tar_resources_aws()`, then they *would* be in memory for the length of the pipeline, and they might be sent over a network in some circumstances. Secret keys should always stay in a safe place on the local machine, and `targets` will never include them in target objects.

> Would you mind pointing me to a resource on how to run the pipeline in different stages with different environment variables?

What I had in mind is something like this:

```r
tar_make(
  names = -any_of("target_with_different_bucket"),
  callr_arguments = list(
    env = c(AWS_ACCESS_KEY_ID = "access_key_1", AWS_SECRET_ACCESS_KEY = "secret_key_1")
  )
)

tar_make(
  names = all_of("target_with_different_bucket"),
  callr_arguments = list(
    env = c(AWS_ACCESS_KEY_ID = "access_key_2", AWS_SECRET_ACCESS_KEY = "secret_key_2")
  )
)
```

Either that or programmatically edit `.Renviron` between stages somehow.
